src/telemetry/metrics/process.rs

```diff
@@ -69,91 +56,128 @@ mod imp {
+            let clock_ticks_per_sec = Self::sysconf(libc::_SC_CLK_TCK, "clock ticks per second")?;
+            Ok(Self {
                 page_size,
                 clock_ticks_per_sec,
             })
         }
 
-        pub fn metrics(&self) -> io::Result<ProcessMetrics> {
-            // XXX potentially blocking call
-            let stat = pid::stat_self()?;
+        fn open_fds(pid: pid_t) -> io::Result<Gauge> {
+            let mut open = 0;
+            for f in fs::read_dir(format!("/proc/{}/fd", pid))? {
+                if !f?.file_type()?.is_dir() {
+                    open += 1;
+                }
+            }
+            Ok(Gauge::from(open))
+        }
 
-            let clock_ticks = stat.utime as u64 + stat.stime as u64;
-            let cpu_seconds_total = Counter::from(clock_ticks / self.clock_ticks_per_sec);
-            let virtual_memory_bytes = Gauge::from(stat.vsize as u64);
-            let resident_memory_bytes = Gauge::from(stat.rss as u64 * self.page_size);
-
-            let metrics = ProcessMetrics {
-                cpu_seconds_total,
-                virtual_memory_bytes,
-                resident_memory_bytes,
-                open_fds: open_fds(stat.pid)?,
-                max_fds: max_fds()?,
-            };
+        fn max_fds() -> io::Result<Option<Gauge>> {
+            let limit = pid::limits_self()?.max_open_files;
+            let max_fds = limit.soft.or(limit.hard)
+                .map(|max| Gauge::from(max as u64));
+            Ok(max_fds)
+        }
 
-            Ok(metrics)
+        fn sysconf(num: libc::c_int, name: &'static str) -> Result<u64, io::Error> {
+            match unsafe { libc::sysconf(num) } {
+                e if e <= 0 => {
+                    let error = io::Error::last_os_error();
+                    error!("error getting {}: {:?}", name, error);
+                    Err(error)
+                },
+                val => Ok(val as u64),
+            }
         }
     }
 
+    impl fmt::Display for System {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            // XXX potentially blocking call
+            let stat = match pid::stat_self() {
+                Ok(stat) => stat,
+                Err(err) => {
+                    warn!("failed to read process stats: {}", err);
+                    return Ok(());
+                }
+            };
 
-    fn open_fds(pid: pid_t) -> io::Result<Gauge> {
-        let mut open = 0;
-        for f in fs::read_dir(format!("/proc/{}/fd", pid))? {
-            if !f?.file_type()?.is_dir() {
-                open += 1;
+            let clock_ticks = stat.utime as u64 + stat.stime as u64;
+            Self::process_cpu_seconds_total.fmt_help(f)?;
+            Self::process_cpu_seconds_total.fmt_metric(
+                f,
+                Counter::from(clock_ticks / self.clock_ticks_per_sec),
+            )?;
+
+            if let Ok(open_fds) = Self::open_fds(stat.pid) {
```

Comment:
Is the error case not a big deal, and this it's ignored? Or should it be logged?